[BUG] openAPI auth bearer auth issue

### Checklist

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq.html#frequently-asked-questions) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)
- [x] I am using the latest version of Schemathesis

### Describe the bug

While using an openAPI schema with a securityscheme of type bearer: https://swagger.io/docs/specification/authentication/bearer-authentication/ I get the following error:

```
E       schemathesis.exceptions.CheckFailed:
E
E       1. Authentication declared but not enforced for this operation
E
E           The API returned `200 OK` for `GET /<PATH>` with invalid auth.
E
E       [200] OK:
E
E           `[]`
E
E       Reproduce with:
E
E           curl -X GET -H 'Authorization: [Filtered]' http://localhost/api/dashboard/trending-clusters
```

The authentication is however valid (therefore the 200 response)

The error is triggered by this codeblock: https://github.com/schemathesis/schemathesis/blob/master/src/schemathesis/specs/openapi/checks.py#L368C1-L372C1

Only if ignored_auth check is enabled.

This happens because `_contains_auth` returns  `AuthKind.GENERATED` which happens since ctx inside contains auth is always None. Which is as far as I could see a result of the securityscheme type.

It doesn't matter if I pass the authentication directly to call_wsgi() via headrs=<authheader> or if I set it via a @schema.auth() decorator. Both send the correct header and successfully authenticate, but the `ignored_auth` check fails. In the case of directly set headers, because `parameter["name"] in ctx.headers` fails if set via the decorator because ctx is None.

### To Reproduce

🚨 **Mandatory** 🚨: Steps to reproduce the behavior:

1. Run this command:

```python
app = create_app()
schemathesis.experimental.OPEN_API_3_1.enable()
schema = schemathesis.from_wsgi("/api/doc/swagger.json", app, skip_deprecated_operations=True)

@schema.auth()
class UserAuth:
    def get(self, case, context):
        client = werkzeug.Client(context.app)
        response = client.post("/api/auth/login", json={"username": "admin", "password": app.config["PRE_SEED_PASSWORD_ADMIN"]})
        if not response or not response.json:
            raise AssertionError(response.text)
        return response.json["access_token"]

    def set(self, case, data, context):
        case.headers = case.headers or {}
        case.headers["Authorization"] = f"Bearer {data}"



@schema.parametrize(endpoint="^/api/dashboard")
@settings(max_examples=50, suppress_health_check=(HealthCheck.function_scoped_fixture,))
def test_dashboard(case):
    response = case.call_wsgi()
    case.validate_response(
        response,
        checks=(ignored_auth),
    )
```

2. See error

Please include a minimal API schema causing this issue:

See [taranis-ai/openapi3_1.yaml](https://github.com/taranis-ai/taranis-ai/blob/master/src/core/core/static/openapi3_1.yaml) for a more comprehensive example

```yaml
openapi: 3.1.0
security:
  - APIKey: []
  - UserAuth: []

paths:
  /dashboard:
    get:
      description: get trending clusters
      security:
        - UserAuth: []
      responses:
        '200':
          description: success
          content:
            application/json:
              schema:
                type: array
                items:
                  type: object
                  properties:
                    size:
                      type: integer
                    name:
                      type: string
                    tags:
                      type: array
                      items:
                        type: object
        '401':
          $ref: "#/components/responses/401Unauthorized"
        '404':
          $ref: "#/components/responses/404NotFound"
        '500':
          description: Internal server error
 
components:
  securitySchemes:
    APIKey:
      type: http
      in: header
      scheme: bearer
      bearerFormat: APIKEY
    UserAuth:
      type: http
      in: header
      scheme: bearer
      bearerFormat: JWT

  responses:
    404NotFound:
      description: "Not found response"
    401Unauthorized:
      description: authorization error
      content:
        application/json:
          schema:
            type: object
            properties:
              error:
                type: string
    405MethodNotAllowed:
      description: method not allowed

```

### Expected behavior

Since I send a valid authentication header a valid response should be expected.

### Environment

```
- OS: Linux Fedora 40
- Python version: 3.12
- Schemathesis version: 3.36.0
- Spec version: 3.1.0
```

### Additional context

Removing igonred_auth from the checks works as a workaround right now.

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -4,6 +4,10 @@ Changelog
 :version:`Unreleased <v3.36.0...HEAD>` - TBD
 --------------------------------------------
 
+**Fixed**
+
+- ``ignored_auth`` false positives on custom auth. :issue:`2462`
+
 .. _v3.36.0:
 
 :version:`3.36.0 <v3.35.5...v3.36.0>` - 2024-09-15

--- a/src/schemathesis/auths.py
+++ b/src/schemathesis/auths.py
@@ -450,6 +450,7 @@ class AuthStorage(Generic[Auth]):
             data: Auth | None = _provider_get(provider, case, context)
             if data is not None:
                 provider.set(case, data, context)
+                case._has_explicit_auth = True
                 break
 
 

--- a/src/schemathesis/models.py
+++ b/src/schemathesis/models.py
@@ -184,6 +184,7 @@ class Case:
     # The way the case was generated (None for manually crafted ones)
     data_generation_method: DataGenerationMethod | None = None
     _auth: requests.auth.AuthBase | None = None
+    _has_explicit_auth: bool = False
 
     def __repr__(self) -> str:
         parts = [f"{self.__class__.__name__}("]
@@ -562,6 +563,8 @@ class Case:
             body=fast_deepcopy(self.body),
             generation_time=self.generation_time,
             id=self.id,
+            _auth=self._auth,
+            _has_explicit_auth=self._has_explicit_auth,
         )
 
 

--- a/src/schemathesis/specs/openapi/checks.py
+++ b/src/schemathesis/specs/openapi/checks.py
@@ -348,7 +348,7 @@ def ignored_auth(ctx: CheckContext, response: GenericResponse, case: Case) -> bo
     security_parameters = _get_security_parameters(case.operation)
     # Authentication is required for this API operation and response is successful
     if security_parameters and 200 <= response.status_code < 300:
-        auth = _contains_auth(ctx, response.request, security_parameters)
+        auth = _contains_auth(ctx, case, response.request, security_parameters)
         if auth == AuthKind.EXPLICIT:
             # Auth is explicitly set, it is expected to be valid
             # Check if invalid auth will give an error
@@ -412,11 +412,13 @@ def _get_security_parameters(operation: APIOperation) -> list[SecurityParameter]
 
 
 def _contains_auth(
-    ctx: CheckContext, request: PreparedRequest, security_parameters: list[SecurityParameter]
+    ctx: CheckContext, case: Case, request: PreparedRequest, security_parameters: list[SecurityParameter]
 ) -> AuthKind | None:
     """Whether a request has authentication declared in the schema."""
     from requests.cookies import RequestsCookieJar
 
+    if case._has_explicit_auth:
+        return AuthKind.EXPLICIT
     parsed = urlparse(request.url)
     query = parse_qs(parsed.query)  # type: ignore
     # Load the `Cookie` header separately, because it is possible that `request._cookies` and the header are out of sync

--- a/test/specs/openapi/test_ignored_auth.py
+++ b/test/specs/openapi/test_ignored_auth.py
@@ -1,12 +1,13 @@
 import json
 import sys
 from base64 import b64decode
+from unittest.mock import Mock
 
 import pytest
 import requests
-from fastapi import FastAPI, Security, status, HTTPException
+from fastapi import FastAPI, HTTPException, Security, status
 from fastapi.security import APIKeyHeader, HTTPAuthorizationCredentials
-from hypothesis import given, settings
+from hypothesis import Phase, given, settings
 from starlette_testclient import TestClient
 
 import schemathesis
@@ -21,7 +22,10 @@ from schemathesis.specs.openapi.checks import AuthKind, _contains_auth, _remove_
 def run(schema_url, headers=None, **loader_kwargs):
     schema = schemathesis.from_uri(schema_url, **loader_kwargs)
     _, _, _, _, _, _, event, *_ = from_schema(
-        schema, checks=[ignored_auth], headers=headers, hypothesis_settings=settings(max_examples=1)
+        schema,
+        checks=[ignored_auth],
+        headers=headers,
+        hypothesis_settings=settings(max_examples=1, phases=[Phase.generate]),
     ).execute()
     return event
 
@@ -126,7 +130,7 @@ def test_no_failure(schema_url):
 )
 def test_contains_auth(ctx, request_kwargs, parameters, expected):
     request = requests.Request("GET", **request_kwargs).prepare()
-    assert _contains_auth(ctx, request, parameters) == expected
+    assert _contains_auth(ctx, Mock(_has_explicit_auth=False), request, parameters) == expected
 
 
 @pytest.mark.parametrize(
@@ -164,7 +168,7 @@ def test_proper_session(ignores_auth):
     schema = schemathesis.from_asgi("/openapi.json", app)
 
     @given(case=schema["/"]["GET"].as_strategy())
-    @settings(max_examples=10)
+    @settings(max_examples=3, phases=[Phase.generate])
     def test(case):
         client = TestClient(app)
         case.call_and_validate(session=client)
@@ -198,7 +202,7 @@ def test_accepts_any_auth_if_explicit_is_present(ignores_auth):
     schema = schemathesis.from_asgi("/openapi.json", app)
 
     @given(case=schema["/"]["GET"].as_strategy())
-    @settings(max_examples=10)
+    @settings(max_examples=3, phases=[Phase.generate])
     def test(case):
         client = TestClient(app)
         case.call_and_validate(session=client, headers={"x-api-key": "INCORRECT"})
@@ -231,3 +235,53 @@ def test_wsgi(wsgi_app_schema, with_generated, headers):
     else:
         assert "Authorization" not in check.request.headers
         assert json.loads(b64decode(check.response.body)) == {"has_auth": False}
+
+
+@pytest.mark.openapi_version("3.0")
+@pytest.mark.operations("ignored_auth")
+def test_explicit_auth(wsgi_app_schema):
+    kwargs = {"auth": ("foo", "bar")}
+    _, _, _, _, _, _, event, *_ = from_schema(
+        wsgi_app_schema, checks=[ignored_auth], hypothesis_settings=settings(max_examples=1), **kwargs
+    ).execute()
+    check = event.result.checks[-1]
+    assert check.value == Status.failure
+    assert check.name == "ignored_auth"
+
+
+@pytest.mark.skipif(sys.version_info < (3, 10), reason="Typing syntax is not supported on Python 3.9 and below")
+def test_custom_auth():
+    app = FastAPI()
+    token = "TEST"
+
+    @app.get("/", responses={200: {"model": {}}, 401: {"model": {}}, 403: {"model": {}}})
+    async def root(
+        credentials: HTTPAuthorizationCredentials = Security(APIKeyHeader(name="x-api-key")),
+    ):
+        if credentials != token:
+            raise HTTPException(
+                status_code=status.HTTP_401_UNAUTHORIZED,
+                detail="Invalid token",
+            )
+        return {"message": "Hello world"}
+
+    schemathesis.experimental.OPEN_API_3_1.enable()
+
+    schema = schemathesis.from_asgi("/openapi.json", app)
+
+    @schema.auth()
+    class Auth:
+        def get(self, case, context):
+            return token
+
+        def set(self, case, data, context):
+            case.headers = case.headers or {}
+            case.headers["x-api-key"] = data
+
+    @given(case=schema["/"]["GET"].as_strategy())
+    @settings(max_examples=10)
+    def test(case):
+        client = TestClient(app)
+        case.call_and_validate(session=client)
+
+    test()

